include/wx/valtext.h

```diff
@@ -63,43 +65,160 @@ class WXDLLIMPEXP_CORE wxTextValidator: public wxValidator
+    }
+
+    // Helper used by IsIncluded() and IsExcluded() methods
+    inline bool IsStringFound(const wxArrayString& arr,
+                              const wxString& str,
+                              const bool excludeLastString) const
+    {
+        const int idx = arr.Index(str);
+
+        const bool found = idx != wxNOT_FOUND;
+
+        if ( excludeLastString )
+        {
+            return found &&
+                   (int)arr.GetCount() > idx+1; // the [inc|exc]lude char list
+        }
+
+        return found;
+    }
+
+    bool IsIncluded(const wxString& str) const
+    {
+        return IsStringFound(m_includes, str, HasFlag(wxFILTER_INCLUDE_CHAR_LIST));
+    }
+
+    bool IsExcluded(const wxString& str) const
+    {
+        return IsStringFound(m_excludes, str, HasFlag(wxFILTER_EXCLUDE_CHAR_LIST));
+    }
+
+    virtual void DoSetIncludes(const wxArrayString& strs);
+    virtual void DoSetExcludes(const wxArrayString& strs);
+
+    virtual void DoAddInclude(const wxString& str);
+    virtual void DoAddExclude(const wxString& str);
+
+    // returns the error message if the contents of 'str' are invalid
+    wxString IsValid(const wxString& str) const;
+
+    // returns false if the character is invalid
+    virtual bool IsValid(const wxUniChar& c) const = 0;
+
+    // Called by Validate() to do the actual validation
+    virtual wxString DoValidate(const wxString& str) = 0;
+
+    // For each derived class set m_style properly
+    virtual void DoSetStyle(long style){ m_style = style; }
 
 protected:
+    long      m_style;
+    wxString* m_string;
 
-    // returns true if all characters of the given string are present in m_includes
-    bool ContainsOnlyIncludedCharacters(const wxString& val) const;
+    wxArrayString m_includes;
+    wxArrayString m_excludes;
 
-    // returns true if at least one character of the given string is present in m_excludes
-    bool ContainsExcludedCharacters(const wxString& val) const;
+private:
+    wxDECLARE_NO_ASSIGN_CLASS(wxTextValidatorBase);
+    wxDECLARE_ABSTRACT_CLASS(wxTextValidatorBase);
+    wxDECLARE_EVENT_TABLE();
+};
 
-    // returns the error message if the contents of 'val' are invalid
-    virtual wxString IsValid(const wxString& val) const;
+class WXDLLIMPEXP_CORE wxTextValidator: public wxTextValidatorBase
+{
+public:
+    explicit wxTextValidator(wxString *str, long style = wxFILTER_NONE);
+    wxTextValidator(long style = wxFILTER_NONE, wxString *str = NULL);
+
+    wxTextValidator(const wxTextValidator& val);
+
+    virtual ~wxTextValidator(){}
+
+    // Make a clone of this validator (or return NULL) - currently necessary
+    // if you're passing a reference to a validator.
+    // Another possibility is to always pass a pointer to a new validator
+    // (so the calling code can use a copy constructor of the relevant class).
+    virtual wxObject *Clone() const wxOVERRIDE { return new wxTextValidator(*this); }
+    bool Copy(const wxTextValidator& val);
```

Comment:
This can't be right, you're hiding the base class function with this one, which does nothing (see the implementation). I don't know what the intention was, but AFAICS it should simply be removed.